Add RISC-V 64 Support for LuaJIT v2.1

IgnotaYun opened the pull request:
This set of patches adds support for RV64G LP64D on Linux, with full interpreter and JIT support.

Patch ported from PLCT LuaJIT, tested with LuaJIT/LuaJIT-test-cleanup.

See also upstream pr LuaJIT/LuaJIT#1267

IgnotaYun commented:
Oops, seems we have some issues with amd64 target……

IgnotaYun commented:
> Oops, seems we have some issues with amd64 target……

Should have been fixed.

mengzhuo commented:
Can we rerun Travis CI?
Alpine using this package for LuaJit2

IgnotaYun commented:
> Can we rerun Travis CI?
> 
> Alpine using this package for LuaJit2

I'll do a patch refresh soon.

Btw, I'm considering dropping DynASM `call` pseudo in favor of "PLT inline" as it's more performant and less hacky.

IgnotaYun commented:
> Can we rerun Travis CI? Alpine using this package for LuaJit2

Done, all passing.

mengzhuo commented:
@zhuizhuhaomeng PTAL, thanks

yuzibo commented:
Thanks for the work!

Debian luajit has [supported](https://tracker.debian.org/news/1586982/accepted-luajit-210openresty20240815-12-source-into-unstable/) riscv64 based on the patch. This fixed almost >[30](https://gist.github.com/yuzibo/5363a4e6c84d49c3517fe7a485f49540) build dependency issues. So far, there have been no reports of build failures or test failures.

BKPepe commented:
@zhuizhuhaomeng Could you please take a look? 😇

glaubitz commented:
Is there a chance this can be reviewed?

IgnotaYun commented:
> Is there a chance this can be reviewed?

Ultimately, it depends on upstream opinion (i.e., Mike).

I am currently employed full-time at PLCT Lab. As Mike focuses on version 3.0, we will ensure that a functional fork for version 2.1 is available.

IgnotaYun committed:
riscv(jit): add mandatory constants

IgnotaYun committed:
riscv(interp): add DWARF info

IgnotaYun committed:
riscv(support): add extension detection

IgnotaYun committed:
riscv(interp): add register definition

IgnotaYun committed:
riscv(dynasm): add RISC-V support

IgnotaYun committed:
riscv(jit): add IR assembler

IgnotaYun committed:
riscv(jit): add insn emitter

IgnotaYun committed:
riscv(interp): strip excessive extended branch (^B+J)

IgnotaYun committed:
riscv(interp): add helper macros and typedefs

IgnotaYun committed:
riscv(ffi): add call convention and support framework

IgnotaYun committed:
riscv(support,linux): make mremap() non-moving due to VA space woes

IgnotaYun committed:
riscv(jit): add hooks in interpreter

IgnotaYun committed:
riscv(misc): add bytecode listing support

IgnotaYun committed:
riscv(jit): add GDBJIT support

IgnotaYun committed:
riscv(misc): add support in Makefile

IgnotaYun committed:
riscv(interp): add frame definition

IgnotaYun committed:
riscv(support,linux): use HWPROBE for ISE detection

Current SIGILL handler appears to have weird issues with libluajit on
some platform. Considering 6.6 kernel is becoming more common, switch
to HWPROBE for better compatibility.

IgnotaYun committed:
riscv(support): add RISC-V 64 arch base definition

IgnotaYun committed:
riscv(support,linux): add Linux specfic icache sync codepath

IgnotaYun committed:
riscv(interp): add VM builder support

Co-authored-by: Heinrich Schuchardt <heinrich.schuchardt@canonical.com>

IgnotaYun committed:
riscv(misc): add disassmbler support

IgnotaYun committed:
riscv(support): add target definition

IgnotaYun committed:
riscv(interp): add base assembly interpreter VM

IgnotaYun commented:
Bumped to LJRV [ce631de](https://github.com/plctlab/LuaJIT/commit/ce631de436c0d231118ec1e0850f6faf224f7b89) w/o [10a7f26](https://github.com/plctlab/LuaJIT/commit/10a7f26a09a91aefe4035f2c82c58a4ece828e76), based-off OpenResty LuaJIT [v2.1-20251030](https://github.com/openresty/luajit2/releases/tag/v2.1-20251030). This should also apply cleanly as-back-to [v2.1-20250117](https://github.com/openresty/luajit2/releases/tag/v2.1-20250117).

The bump contains many bug fixes to FFI and the interpreter; we recommend that distros apply it if possible.

The next label here would likely introduce a breaking change; in such a case, do not apply until I refresh it.

peterwillcn commented:
any news?

IgnotaYun commented:
Bumped to LJRV [0202af5](https://github.com/plctlab/LuaJIT/commit/0202af591a08479e51f8c842a87b6fd198b67946), based-off OpenResty LuaJIT [v2.1-20251229](https://github.com/openresty/luajit2/releases/tag/v2.1-20251229).

This cannot be applied to versions prior to v2.1-20251229. For such usage, refer to the branch [v2.1-20251030-riscv64](https://github.com/infiWang/luajit2/tree/v2.1-20251030-riscv64), which can be applied cleanly from v2.1-20250117 to v2.1-20251030.

IgnotaYun commented:
> any news?

Thanks for your interest! The current situation is that the upstreaming process is largely tied to Mike Pall's availability and sponsorship.

OpenResty maintainers have indicated they would prefer the patch to land upstream first. Since Mike currently lacks sponsorship for the RISC-V port, progress is naturally slower. While ISCAS is coordinating with RISE for potential funding, we strongly encourage any commercial entities that need this port to consider sponsoring Mike or OpenResty.

A notable precedent is Linaro's sponsorship of the AArch64 port, and here at OpenResty, IBM's sponsorship of the S390X/PPC64 ports. We (ISCAS) will continue to handle all the heavy lifting regarding code updates and reviews, so sponsors can focus purely on supporting the ecosystem's sustainability.

IgnotaYun commented:
Bumped to LJRV [6fdbb10](https://github.com/plctlab/LuaJIT/commit/6fdbb10441a1c1c6834bd6924dcbf2d5e99f1f74), based-off OpenResty LuaJIT [v2.1-20260114](https://github.com/openresty/luajit2/releases/tag/v2.1-20260114).

This cannot be applied to versions prior to v2.1-20260114. For v2.1-20251229, refer to the branch [v2.1-20251229-riscv64](https://github.com/infiWang/luajit2/tree/v2.1-20251229-riscv64). See the thread/repo for previous versions.

IgnotaYun committed:
riscv(misc): add disassmbler support

IgnotaYun committed:
riscv(interp): add register definition

IgnotaYun committed:
riscv(interp): add DWARF info

IgnotaYun committed:
riscv(jit): add IR assembler

IgnotaYun committed:
riscv(support): add RISC-V 64 arch base definition

IgnotaYun committed:
riscv(dynasm): add RISC-V support

IgnotaYun committed:
riscv(interp): add helper macros and typedefs

IgnotaYun committed:
riscv(jit): add hooks in interpreter

IgnotaYun committed:
riscv(interp): add frame definition

IgnotaYun committed:
riscv(ffi): add call convention and support framework

IgnotaYun committed:
riscv(jit): add GDBJIT support

IgnotaYun committed:
riscv(support): add target definition

IgnotaYun committed:
riscv(jit): add insn emitter

IgnotaYun committed:
riscv(interp): add VM builder support

Co-authored-by: Heinrich Schuchardt <heinrich.schuchardt@canonical.com>

IgnotaYun committed:
riscv(support,linux): make mremap() non-moving due to VA space woes

IgnotaYun committed:
riscv(support): add extension detection

IgnotaYun committed:
riscv(jit): add mandatory constants

IgnotaYun committed:
riscv(support,linux): add Linux specfic icache sync codepath

IgnotaYun committed:
riscv(interp): add base assembly interpreter VM

IgnotaYun committed:
riscv(misc): add support in Makefile

IgnotaYun committed:
riscv(misc): add bytecode listing support
